### PR TITLE
Fix bug in index.union with duplicates and not a subset of each other

### PR DESCRIPTION
**Pull request by phofl**
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

The previous fix missed one case where other has dups which are unique in self

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -2979,8 +2979,8 @@ def _union(self, other: Index, sort):
                 value_list.extend([x for x in rvals if x not in value_set])
                 return Index(value_list)._values  # do type inference here
 
-        elif not other.is_unique and not self.is_unique:
-            # self and other both have duplicates
+        elif not other.is_unique:
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py:2982]: why doesnt the `not self.is_unique` case go through here?

Comment (phofl) [reply pandas/core/indexes/base.py:2982]: self is our basis in this case and we add things from other, which are not in self, so if self has duplicates and other has not, we are fine, but if other has duplicates which are unique in self, the dups are not added below.

Example:

```
self = Index([0, 0, 1])
other = Index([0, 1, 2])
```

-> we add 2 to lvals and get out result, but if

```
self = Index([0, 1, 2])
other = Index([0, 0, 1])
```

Nothing is missing in this case, so we add nothing to self, but we want to add one 0.

Sorry is tricky to explain, was this clear enough to get an idea why we are doing this? We could in theory send it through I think, but I believe the function is slower than the code below

Comment (jbrockmendel) [reply pandas/core/indexes/base.py:2982]: makes sense, thanks

**Review by jbrockmendel — APPROVED** [review]
LGTM

**Review by jreback — CHANGES_REQUESTED** [review]
great, can you add this PR number to the original whatsnew entry. ping on green.

**Comment by phofl** [comment]
@jreback greenish, failure unrelated

**Comment by jreback** [comment]
very nice @phofl

**Pull request merged**